src/firmware/uapi/guest/types.rs

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use bitfield::bitfield;
+
+pub(crate) use crate::firmware::linux::guest::types::{
+    SnpDerivedKeyReq, SnpDerivedKeyRsp, SnpReportRsp,
+};
+
+pub use crate::firmware::linux::guest::types::{
+    AttestationReport, Signature, SnpExtReportReq, SnpGuestPolicy, SnpPlatformInfo, SnpReportReq,
+    SnpTcbVersion,
+};
+
+/// Structure of required data for fetching the derived key.
+pub struct SnpDerivedKey {
+    /// Selects the root key to derive the key from.
+    /// 0: Indicates VCEK.
+    /// 1: Indicates VMRK.
+    root_key_select: u32,
+
+    /// Reserved, must be zero
+    _reserved_0: u32,
+
+    /// What data will be mixed into the derived key.
+    pub guest_field_select: GuestFieldSelect,
+
+    /// The VMPL to mix into the derived key. Must be greater than or equal
+    /// to the current VMPL.
+    pub vmpl: u32,
+
+    /// The guest SVN to mix into the key. Must not exceed the guest SVN
+    /// provided at launch in the ID block.
+    pub guest_svn: u32,
+
+    /// The TCB version to mix into the derived key. Must not
+    /// exceed CommittedTcb.
+    pub tcb_version: u64,
+}
+
+impl SnpDerivedKey {
+    /// Create a new instance for requesting an SnpDerivedKey.
+    ///
+    /// # Arguments:
+    ///
+    /// * `root_key_select` - bool - true: use VMRK to derive key, false: use VCEK to derive key.
+    /// * `guest_field_select` - [`GuestFieldSelect`] - Bitfield expressing which will be mixed into the derived key.
+    /// * `vmpl` - u32 - What VMPL level the derived key should be associated with.
+    /// * `guest_svn` - u32 - Guest SVN to mix into the key.
+    /// * `tcb_version` - u64 - The TCB Version to mix into the derived key. Must __NOT__ Exceed Committed TCB.
+    pub fn new(
+        root_key_select: bool,
+        guest_field_select: GuestFieldSelect,
+        vmpl: u32,
+        guest_svn: u32,
+        tcb_version: u64,
+    ) -> Self {
+        Self {
+            root_key_select: u32::from(root_key_select),
+            _reserved_0: Default::default(),
+            guest_field_select,
+            vmpl,
+            guest_svn,
+            tcb_version,
+        }
+    }
+
+    /// Obtain a copy of the root key select value (Private Field)
+    pub fn get_root_key_select(&self) -> u32 {
+        self.root_key_select
+    }
+}
+
+/// A raw representation of the PSP Report Response after calling SNP_GET_DERIVED_KEY.
+pub struct DerivedKeyRsp {
+    /// The status of key derivation operation.
+    /// 0h: Success.
+    /// 16h: Invalid parameters
+    pub status: u32,
+
+    _reserved_0: [u8; 28],
+
+    /// The requested derived key if [`DerivedKeyRsp::status`] is 0h.
+    pub key: [u8; 32],
```

Comment:
Is `status` required here? Rather, if status shows to be invalid, we can return an `Error` from the user API, and if not, return an `Ok(DerivedKey)`.... just a thought for later on. This doesn't need to be addressed in this PR.